arangod/IResearch/IResearchLink.cpp

```diff
@@ -521,99 +538,69 @@ arangodb::Result IResearchLink::commitUnsafe() {
   if (!engine) {
     return {
       TRI_ERROR_INTERNAL,
-      "failure to get storage engine while committing arangosearch link '" + std::to_string(id()) + "'"};
+      "failure to get storage engine while committing arangosearch link '" + std::to_string(id()) + "'"
+    };
+  }
+
+  auto subscription = std::static_pointer_cast<IResearchFlushSubscription>(_flushSubscription);
+
+  if (!subscription) {
+    // already released
+    return { };
   }
 
   try {
-    // upcoming 'index_writer::commit()' will wait until all 'document_context's
-    // held by any transaction will be released, so 'tick' value can be less
-    // than actual engine tick when iresearch commit happens
-    auto const tick = engine->currentTick();
+    auto lastCommittedTick = engine->currentTick();
 
-    _dataStore._writer->commit();
+    _dataStore._writer->commit(_before_commit);
 
     SCOPED_LOCK(_readerMutex);
     auto reader = _dataStore._reader.reopen(); // update reader
 
     if (!reader) {
       // nothing more to do
-      LOG_TOPIC("37bcf", WARN, arangodb::iresearch::TOPIC)
+      LOG_TOPIC("37bcf", WARN, iresearch::TOPIC)
         << "failed to update snapshot after commit, run id '" << size_t(&runId)
         << "', reuse the existing snapshot for arangosearch link '" << id() << "'";
 
       return {};
     }
 
-    if (_dataStore._reader == reader
-        && _dataStore._recovery_range_start == reader.meta().filename) {
+    if (_dataStore._reader == reader) {
+      LOG_TOPIC("7e319", TRACE, iresearch::TOPIC)
+        << "no changes registered for arangosearch link '" << id()
+        << "' got last operation tick '" << lastCommittedTick
+        << "', run id '" << size_t(&runId) << "'";
 
-      // reader not modified
-      if (_flushCallback) {
-        // upgrade tick without writing WAL entry
-        return _flushCallback(VPackSlice::noneSlice(), tick);
-      }
+      // no changes, can release the latest tick before commit
+      subscription->tick(lastCommittedTick);
 
       return {};
     }
 
-    // if WAL 'Flush' recovery is enabled (must be for recoverable DB scenarios)
-    if (_flushCallback && RecoveryState::DONE == _dataStore._recovery) {
-      auto& checkpoint = reader.meta().filename;
-      auto checkpointFile = checkpoint + std::string(IRESEARCH_CHECKPOINT_SUFFIX);
-      auto ref = irs::directory_utils::reference(*(_dataStore._directory), checkpointFile, true);
-      arangodb::velocypack::Builder builder;
+    // update reader
+    _dataStore._reader = reader;
 
-      builder.add(arangodb::velocypack::Value(checkpoint));
+    // invalidate query cache
+    aql::QueryCache::instance()->invalidate(&(_collection.vocbase()), _viewGuid);
```

Comment:
why is the query cache invalidated in here ?

Reply:
is this not beeing done in `TransactionState::clearQueryCache()` ?

Reply:
ArangoSearch has a number of dedicated maintenance jobs which can upgrade snapshot asynchronously. That's why we need to invalidate query cache here as well.